### PR TITLE
Laboratory work #2, Krinochkina Darya - 20FPL1

### PR DESCRIPTION
**Pull request by DashaKrinochkina**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
push some changes, please; deadline is coming

File: lab_2/main.py [hunk 1]
```diff
@@ -13,7 +13,16 @@ def get_freq_dict(tokens: list) -> dict or None:
...
     """
-    pass
+    if not isinstance(tokens, list):
+        return None
+    for i in tokens:
+        if not i:
+            return None
+
+    freq_dict = {}
+    for i in tokens:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:23]: better to use names like token/word/etc

File: lab_2/main.py [hunk 2]
```diff
@@ -13,7 +13,16 @@ def get_freq_dict(tokens: list) -> dict or None:
...
+        freq_dict[i] = round(tokens.count(i)/len(tokens), 5)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:24]: what happens here?

File: lab_2/main.py [hunk 3]
```diff
@@ -24,7 +33,19 @@ def get_language_profiles(texts_corpus: list, language_labels: list) -> dict or
...
+    for i in texts_corpus:
+        if not i:
+            return None
+    for i in language_labels:
+        if not i:
+            return None
+
+    language_profiles = {}
+    for i in range(len(language_labels)):
+        language_profiles[language_labels[i]] = get_freq_dict(texts_corpus[i])
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:47]: what happens here?

File: lab_2/main.py [hunk 4]
```diff
@@ -33,7 +54,14 @@ def get_language_features(language_profiles: dict) -> list or None:
         and sorts them in alphabetical order
     :param language_profiles: a dictionary of dictionaries - language profiles
     """
-    pass
+    if not isinstance(language_profiles, dict) or not language_profiles:
+        return None
+
+    features = []
+    for i in language_profiles.values():
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:61]: same about naming

File: lab_2/main.py [hunk 5]
```diff
@@ -33,7 +54,14 @@ def get_language_features(language_profiles: dict) -> list or None:
...
+        for key in i.keys():
+            features.append(key)
+    return sorted(features)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:64]: what happens in this function?

File: lab_2/main.py [hunk 6]
```diff
@@ -43,7 +71,18 @@ def get_text_vector(original_text: list, language_profiles: dict) -> list or Non
...
     :param language_profiles: a dictionary of dictionaries - language profiles
     """
-    pass
+    if not isinstance(language_profiles, dict) or not isinstance(original_text, list):
+        return None
+
+    text_vector = []
+    for word in get_language_features(language_profiles):
+        if word in original_text:
+            for d in language_profiles:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:80]: naming

File: lab_2/main.py [hunk 7]
```diff
@@ -43,7 +71,18 @@ def get_text_vector(original_text: list, language_profiles: dict) -> list or Non
...
+                if word in language_profiles[d]:
+                    text_vector.append(language_profiles[d][word])
+        else:
+            text_vector.append(0)
+    return text_vector
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:85]: what happens in this function?

File: lab_2/main.py (hunk 1 above)
Comment (DashaKrinochkina) [reply lab_2/main.py]: done

File: lab_2/main.py (hunk 3 above)
Comment (DashaKrinochkina) [reply lab_2/main.py:47]: I get language label from the labels list by index. Then I get text from corpus by the same index and give it to function.
Then I create a language profiles by adding label as a key and the result of function(frequency dictionary) as a value

File: lab_2/main.py (hunk 6 above)
Comment (DashaKrinochkina) [reply lab_2/main.py]: done

File: lab_2/main.py (hunk 7 above)
Comment (DashaKrinochkina) [reply lab_2/main.py:85]: This fuction returns frequency vector for a text. If the uniqe word from features list is in text, its frequency was added to vector list, if not, zero was added

File: lab_2/main.py [hunk 8]
```diff
@@ -128,7 +174,34 @@ def predict_language_knn(unknown_text_vector: list, known_text_vectors: list,
...
+
+    distances = []
+    if metric == 'manhattan':
+        for vector in known_text_vectors:
+            distances.append(calculate_distance_manhattan(unknown_text_vector, vector))
+    else:
+        for vector in known_text_vectors:
+            distances.append(calculate_distance(unknown_text_vector, vector))
+
+    k_nearest = sorted(zip(language_labels, distances), key=lambda x: x[1])[:k]
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:197]: what is lambda and what does it do here?

File: lab_2/main.py [hunk 9]
```diff
@@ -128,7 +174,34 @@ def predict_language_knn(unknown_text_vector: list, known_text_vectors: list,
...
+    language_freq = {}
+    for language in k_nearest:
+        if language[0] not in language_freq:
+            language_freq[language[0]] = 1
+        language_freq[language[0]] += 1
+    possible_language = [max(language_freq, key=language_freq.get), min(distances)]
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:203]: what happens here?

File: lab_2/main.py [hunk 10]
```diff
@@ -139,7 +212,22 @@ def get_sparse_vector(original_text: list, language_profiles: dict) -> list or N
...
+            for dictionary in language_profiles:
+                if word in language_profiles[dictionary]:
+                    text_vector.append(language_profiles[dictionary][word])
+        else:
+            text_vector.append(0)
+    updated_vector = []
+    for i, number in enumerate(text_vector):
+        if number > 0:
+            updated_vector.append([i, number])
+    return updated_vector
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:230]: can you tell me what happens in this function?

File: lab_2/main.py [hunk 11]
```diff
@@ -149,7 +237,27 @@ def calculate_distance_sparse(unknown_text_vector: list,
...
+    for pare in known_text_vector:
+        distance_dict[pare[0]] = pare[1]
+    for pare in unknown_text_vector:
+        if pare[0] in distance_dict.keys():
+            distance_dict[pare[0]] -= pare[1]
+        else:
+            distance_dict[pare[0]] = pare[1]
+    for value in distance_dict.values():
+        distance_sum += value ** 2
+    return round(distance_sum ** 0.5, 5)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:260]: and here too?

File: lab_2/start.py [hunk 12]
```diff
@@ -37,7 +39,30 @@
...
+    language_profiles = get_language_profiles(text_corpus, language_labels)
+    known_text_vectors = [get_sparse_vector(text, language_profiles) for text in text_corpus]
+    k = 3
+    RESULT = []
+    for text in UNKNOWN_SAMPLES:
+        unknown_text = remove_stop_words(tokenize(text), stop_words)
+        unknown_text_vector = get_sparse_vector(unknown_text, language_profiles)
+        prediction = predict_language_knn_sparse(unknown_text_vector, known_text_vectors,
+                                                 language_labels, k)
+        RESULT.append(prediction[0])
```
Comment (ValeriyaKuznetsova) [inline lab_2/start.py:64]: what happens in `start.py`?

Comment (ValeriyaKuznetsova) [reply lab_2/start.py:64]: let's predict only one language here

File: lab_2/main.py (hunk 11 above)
Comment (DashaKrinochkina) [reply lab_2/main.py:266]: this function calculates distance for sparsed vecrtor. firstly it adds all pares index(key)-frequency(value) from known_text_vector to dictionary, then it stars to add pares from unknown_text_vector, but if pares from unknown and known vectors contain frequencies with same indexes, funtion adds index and difference between frequencies. then it calulates sum of squared values and extracts square root

**Pull request closed without merging**